src/elbo.jl

```diff
@@ -0,0 +1,148 @@
+    scale = n_data / n_batch
+    return sum(variational_exp) * scale - kl_term
+end
+
+
+"""
+    elbo(fx::LatentFiniteGP, y::AbstractVector, fz::FiniteGP, q::MvNormal; n_data=length(y))
+
+Compute the ELBO for a LatentGP with a possibly non-conjugate likelihood.
+"""
+function elbo(
+    lfx::LatentFiniteGP,
+    y::AbstractVector,
+    fz::FiniteGP,
+    q::AbstractMvNormal;
+    n_data=length(y)
+)
+    n_batch = length(y)
+    kl_term, f_mean, f_var = _elbo_intermediates(lfx.fx, fz, q)
+    
+    variational_exp = expected_loglik(y, f_mean, f_var, lfx.lik)
+    scale = n_data / n_batch
+    return variational_exp * scale - kl_term
+end
+
+# Computes the common intermediates needed for the ELBO
+function _elbo_intermediates(
+    fx::FiniteGP,
+    fz::FiniteGP,
+    q::AbstractMvNormal
+)
+    kl_term = kl_divergence(q, fz)
+    post = approx_posterior(SVGP(), fz, q)
+    f_mean, f_var = mean_and_var(post, fx.x)
+    return kl_term, f_mean, f_var
+end
+
+"Likelihoods which take a scalar (or vector of scalars) as input and return a single scalar."
+ScalarLikelihood = Union{BernoulliLikelihood,PoissonLikelihood}
+
+"""
+    expected_loglik(y, f_mean, f_var, [Σy | lik])
+
+This function computes the expected log likelihood:
+
+```math
+    ∫ q(f) log p(y | f) df
+```
+where `p(y | f)` is the process likelihood.
+
+`q(f)` is an approximation to the latent function values `f` given by:
+```math
+    q(f) = ∫ p(f | u) q(u) du
+```
+where `q(u)` is the variational distribution over inducing points (see
+[`elbo`](@ref)).
+
+Where possible, this expectation is calculated in closed form. Otherwise, it is
+approximated using Gauss-Hermite quadrature by default.
+
+# Extended help
+
+`q(f)` is assumed to be an `MvNormal` distribution and `p(y | f)` is assumed to
+have independent marginals such that only the marginals of `q(f)` are required.
+"""
+
+function expected_loglik end
+
+"""
+    expected_loglik(y::AbstractVector{<:Real}, f_mean::AbstractVector, f_var::AbstractVector, Σy::AbstractVector)
+
+The expected log likelihood for a Gaussian likelihood, computed in closed form.
+"""
+# The closed form expected loglikelihood for a Gaussian likelihood
+function expected_loglik(
+    y::AbstractVector{<:Real},
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    Σy::AbstractVector
+)
+    return sum(-0.5 * (log(2π) .+ log.(Σy) .+ ((y .- f_mean).^2 .+ f_var) ./ Σy))
+end
+
+"""
+    expected_loglik(y::AbstractVector, f_mean::AbstractVector, f_var::AbstractVector, lik::ScalarLikelihood; n_points=20)
+
+The expected log likelihood for a `ScalarLikelihood`, approximated via
+Gauss-Hermite quadrature with `n_points` quadrature points.
+"""
+
+function expected_loglik(
+    y::AbstractVector,
+    f_mean::AbstractVector,
+    f_var::AbstractVector,
+    lik::ScalarLikelihood;
+    n_points=20
+)
+    return sum(gauss_hermite_quadrature(y, f_mean, f_var, lik; n_points=n_points))
+end
+
+function kl_divergence(q::AbstractMvNormal, p::AbstractMvNormal)
```

Comment:
```suggestion
function StatsBase.kldivergence(q::AbstractMvNormal, p::AbstractMvNormal)
```
(and changing all calls from `kl_divergence` to `kldivergence`, of course)

I was expecting to see implementations from Distributions.jl already, but all I could find was https://github.com/theogf/KLDivergences.jl - would it be worth building on that @theogf ?

Reply:
Well I started building this to take it out of AGP.jl, but it is quite "brittle"

Reply:
Ah, I didn't see that package - I suppose I'll keep the implementation here for now, and replace it if it gets more stable. @theogf what makes it brittle atm? I'd be up for working on it if needed.

Reply:
> what makes it brittle atm

I don't think it's very well tested and does not account for edge cases...

Let me have a look back to what I did, for example I did not know StatsBase.kldivergence existed, I should probably reuse that!

Reply:
Have now switched to using `KLDivergences.jl` for this